Visitor can have the option to create an article WITH CATEGORIES

boamatule opened the pull request:
# [PT Story](https://www.pivotaltracker.com/story/show/166152109)

## Changes proposed in this pull request:
* Adds Unit Test for the Category model
* Adds Category models
* Adds Associations between Articles and Categories
* Adds Category Controller
* Adds Category to routes
* Adds Seed to populate the Category table


## What I have learned working on this feature:
* Handling Associations

boamatule committed:
Adds test for the model category.

stefankarlberg committed:
Adds category model with column name. Adds validation for column name to the model.

boamatule committed:
Adds new column category_id to articles table as foreign key.

stefankarlberg committed:
Adds new param category_id in article controller. Adds associations in models. Adds adjustment in article API unit test to match the new assocoation. Adds new FactoryBot to category.

stefankarlberg committed:
Adds category_id to serializer, adds categories to routes, adds seeds for autopopulate the categories in db, adds category controller and creates request tests for categories API.

raptorf1 commented on app/controllers/api/v1/category_controller.rb:
```diff
@@ -0,0 +1,6 @@
+class CategoryController < ApplicationController
```
do we need 2 controllers (`category` and `categories`)????

raptorf1 commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,14 @@
 class ApplicationController < ActionController::API
```
why are we putting code in the `ApplicationController`?

raptorf1 commented on db/migrate/20190525141132_category_id_to_articles.rb:
```diff
@@ -0,0 +1,5 @@
+class CategoryIdToArticles < ActiveRecord::Migration[5.2]
```
the same thing is reffered on the first migration file also. any reason for that?

raptorf1 commented on spec/requests/api/v1/articles_spec.rb:
```diff
@@ -5,7 +5,8 @@
   describe 'GET /api/v1/articles' do
     before do
-      5.times { FactoryBot.create(:article) }
+      category = create(:category)
```
if the Factory creates articles with categories, do we need that line?

zanenkn commented on app/controllers/api/v1/categories_controller.rb:
```diff
@@ -0,0 +1,28 @@
+      render status: 422, json: { error: category.errors.full_messages }
+    end
+  end
+
```
will we be creating new categories? imo unnecessary method.

stefankarlberg committed:
Adjusts test, adjusts name in FatcoryBot.

zanenkn commented on config/routes.rb:
```diff
@@ -4,7 +4,7 @@
     namespace :v1 do
-      resources :articles, only: [:index, :create, :show]
+      resources :categories, :articles, only: [:index, :create, :show]
     end
```
id say we should limit the routes open for the `:categories`, for example we wont be editing or deleting categories so those should definitely be excluded.

zanenkn commented on db/seeds.rb:
```diff
@@ -1,7 +1,42 @@
+  [ "Food" ],
+  [ "Travel" ],
+  [ "Real State" ]
+]
```
"Real Estate" i guess.

oliverochman replied:
please update the "Real State" string

raptorf1 replied:
done

boamatule committed:
Merge branch 'create_an_article_with_categories' of https://github.com/stefankarlberg/Glocal_news-API into create_an_article_with_categories

boamatule committed:
Adds serializer for Categories;
Removed methods in category controler;
Removed posts tests in resquest categories;
All green.

tochman reviewed (changes requested):
I can't really see how this constitutes a feature nor how it fulfill the request implied in the PT title/user story. Also, please make sure to name your PR's the same as the corresponding PT story.

tochman commented on app/controllers/api/v1/categories_controller.rb:
```diff
@@ -0,0 +1,7 @@
+
+  def index
+    categories = Category.all
+    render json: { data: categories }, each_serializer: Categories::IndexSerializer
```
Why do you use `data` as a key name? Why not use the resource name?

boamatule replied:
We just changed the name to categories which is the right resource name.

tochman commented on spec/models/article_spec.rb:
```diff
@@ -1,3 +1,5 @@
+require 'rails_helper'
```
Are you sure you need this require? Have you checked the other specs? What are you including in your `.rspec`?

boamatule replied:
We saw that we have the `require rails_helper` code  included in the .rspec. 
Removed now.

boamatule committed:
removed 'require rails_helper' in the request tests.

boamatule committed:
changes the resourse name in the category.

oliverochman commented on spec/models/article_spec.rb:
```diff
@@ -14,11 +17,12 @@
+    it { is_expected.to validate_presence_of(:category_id) }
   end
-
+  
```
You need to test the association between article and category here

raptorf1 replied:
done

boamatule committed:
Addresses the comments from the PR;
Solving merging conflicts.

raptorf1 committed:
Fixes tests after merge conflicts. Edits index action of Articles controller cause the serializer was not working previously.

raptorf1 commented:
> I can't really see how this constitutes a feature nor how it fulfill the request implied in the PT title/user story. Also, please make sure to name your PR's the same as the corresponding PT story.

User story description changed. It is marked as an API story only. There is another one for the CLIENT side.

raptorf1 committed:
Fixes serializer issue for the Categories controller also

raptorf1 committed:
Removes unused routes for categories. Removes unused argument from articles factory.

leiter007 reviewed (commented):
Looks great, good adjustments after latest comments.

raptorf1 committed:
Configures serializers so that every articles GET request contains the category as a response
